btree/btree.c

```diff
@@ -0,0 +1,13098 @@
+	do {                                                                 \
+		*(cred) = M0_BE_TX_CREDIT(50, (4 * 1024));            \
+	} while(0)
+
+#endif
+
+/**
+ *  --------------------------------------------
+ *  Section END - BTree Structure and Operations
+ *  --------------------------------------------
+ */
+
+/**
+ *  ---------------------------------------------------
+ *  Section START - BTree Node Structure and Operations
+ *  ---------------------------------------------------
+ */
+
+
+/**
+ * "Address" of a node in a segment.
+ *
+ * Highest 8 bits (56--63) are reserved and must be 0.
+ *
+ * Lowest 4 bits (0--3) contains the node size, see below.
+ *
+ * Next 5 bits (4--8) are reserved and must be 0.
+ *
+ * Remaining 47 (9--55) bits contain the address in the segment, measured in 512
+ * byte units.
+ *
+ * @verbatim
+ *
+ *  6      5 5                                            0 0   0 0  0
+ *  3      6 5                                            9 8   4 3  0
+ * +--------+----------------------------------------------+-----+----+
+ * |   0    |                     ADDR                     |  0  | X  |
+ * +--------+----------------------------------------------+-----+----+
+ *
+ * @endverbatim
+ *
+ * Node size is 2^(9+X) bytes (i.e., the smallest node is 512 bytes and the
+ * largest node is 2^(9+15) == 16MB).
+ *
+ * Node address is ADDR << 9.
+ *
+ * This allows for 128T nodes (2^47) and total of 64PB (2^56) of meta-data per
+ * segment.
+ *
+ * NOTE: Above design is made obsolete from EOS-25149. Now Address of the node
+ * in segment will be stored as it is. The node size will be stored in the
+ * respective node type header.
+ */
+struct segaddr {
+	uint64_t as_core;
+};
+
+enum {
+	NODE_SHIFT_MIN = 9,
+};
+
+static struct segaddr  segaddr_build(const void *addr);
+static void           *segaddr_addr (const struct segaddr *addr);
+static uint32_t        segaddr_ntype_get(const struct segaddr *addr);
+static bool            segaddr_header_isvalid(const struct segaddr *addr);
+
+/**
+ * B-tree node in a segment.
+ *
+ * This definition is private to the node sub-module.
+ */
+struct segnode;
+
+/**
+ * Tree descriptor.
+ *
+ * A tree descriptor is allocated for each b-tree actively used by the b-tree
+ * module.
+ */
+struct td {
+	const struct m0_btree_type *t_type;
+
+	/**
+	 * The lock that protects the fields below t_lock. The fields above are
+	 * read-only after the tree root is loaded into memory.
+	 */
+	struct m0_rwlock            t_lock;
+	struct nd                  *t_root;   /** Points to root node on seg */
+	int                         t_height; /** Height of the tree */
+	int                         t_ref;    /** Reference count */
+	struct m0_be_seg           *t_seg;    /** Segment hosting tree nodes. */
+	struct m0_fid               t_fid;    /** Fid of the tree. */
+	struct m0_btree_rec_key_op  t_keycmp; /** User Key compare function */
```

Comment:
Let's postpone key comparators until they are needed. I see at least 2 problems with this feature:

1. How do we encode the key comparison function on the storage? It is desirable that generic utilities (like recovery or tree dumping tool) were able to work with "unknown" trees.
2. There is no, in general, a guarantee that keys are contiguous in the node. That is, m0_btree_rec_key_op::rko_keycmp() should take const struct m0_buf_vec-s (or m0_buf_vec_cursor-s) as parameters, instead of void *. And with discontiguous keys, non-trivial key comparators are tricky.